### PR TITLE
compiletest: clean rmeta data (from "cargo check") before running compiletest

### LINKED ISSUE #2139: mut_mut_macro test fails after 'cargo check': "multiple matching crates for `lazy_static`"
I have a problem with tests execution. Every time, when I do `cargo check`, subsequent launch of `cargo test` fails.

It reproduces like this:
```
cargo clean
cargo test # SUCCEED
cargo test # SUCCEED
cargo check
cargo test # FAILED
```

It worked in such a way for me with different releases.

My current environment:
```
$ rustup show
Default host: x86_64-unknown-linux-gnu

nightly-x86_64-unknown-linux-gnu (default)
rustc 1.22.0-nightly (02a24dbdd 2017-10-13)
```

Failed test:
```
failures:

---- [run-pass] run-pass/mut_mut_macro.rs stdout ----

error: compilation failed!
status: exit code: 101
command: target/debug/clippy-driver tests/run-pass/mut_mut_macro.rs -L target/debug/test_build_base --target=x86_64-unknown-linux-gnu --error-format json -L target/debug/test_build_base/mut_mut_macro.stage-id.run-pass.libaux -C prefer-dynamic -o target/debug/test_build_base/mut_mut_macro.stage-id -L target/debug/ -L target/debug/deps -Dwarnings
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
{"message":"multiple matching crates for `lazy_static`","code":{"code":"E0464","explanation":null},"level":"error","spans":[{"file_name":"tests/run-pass/mut_mut_macro.rs","byte_start":74,"byte_end":99,"line_start":7,"line_end":7,"column_start":1,"column_end":26,"is_primary":true,"text":[{"text":"extern crate lazy_static;","highlight_start":1,"highlight_end":26}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[{"message":"candidates:","code":null,"level":"note","spans":[],"children":[],"rendered":null},{"message":"path: /home/yk/dev/rust-clippy/target/debug/deps/liblazy_static-e94bdb055ffaade6.rlib","code":null,"level":"note","spans":[],"children":[],"rendered":null},{"message":"crate name: lazy_static","code":null,"level":"note","spans":[],"children":[],"rendered":null},{"message":"crate name: lazy_static","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":null}
{"message":"can't find crate for `lazy_static`","code":{"code":"E0463","explanation":"\nA plugin/crate was declared but cannot be found. Erroneous code example:\n\n```compile_fail,E0463\n#![feature(plugin)]\n#![plugin(cookie_monster)] // error: can't find crate for `cookie_monster`\nextern crate cake_is_a_lie; // error: can't find crate for `cake_is_a_lie`\n```\n\nYou need to link your code to the relevant crate in order to be able to use it\n(through Cargo or the `-L` option of rustc example). Plugins are crates as\nwell, and you link to them the same way.\n"},"level":"error","spans":[{"file_name":"tests/run-pass/mut_mut_macro.rs","byte_start":74,"byte_end":99,"line_start":7,"line_end":7,"column_start":1,"column_end":26,"is_primary":true,"text":[{"text":"extern crate lazy_static;","highlight_start":1,"highlight_end":26}],"label":"can't find crate","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:525:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.
thread 'main' panicked at 'rustc_thread failed: Any', /checkout/src/libcore/result.rs:906:4

------------------------------------------

thread '[run-pass] run-pass/mut_mut_macro.rs' panicked at 'explicit panic', /home/yk/.cargo/registry/src/github.com-1ecc6299db9ec823/compiletest_rs-0.2.10/src/runtest.rs:2374:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

Full test log: https://hastebin.com/etekarumis

I don't understand why it happend. Does it reproduce for anybody else?

### LINKED ISSUE #2896: "cargo check"  makes  "cargo test"  fail
repo is at 
30a9879c461ce18ff120e513a2f6a16d0f2f4b80

````
cargo check
[...]
cargo test
[...]
test [run-pass] run-pass/regressions.rs ... ok
test [run-pass] run-pass/used_underscore_binding_macro.rs ... FAILED
test [run-pass] run-pass/returns.rs ... ok
test [run-pass] run-pass/whitelist/conf_whitelisted.rs ... ok
test [run-pass] run-pass/single-match-else.rs ... ok
failures:
---- [run-pass] run-pass/used_underscore_binding_macro.rs stdout ----
error: compilation failed!
status: exit code: 101
command: "target/debug/clippy-driver" "tests/run-pass/used_underscore_binding_macro.rs" "-L" "/home/matthias/vcs/github/rust-clippy/target/debug/test_build_base" "--target=x86_64-unknown-linux-gnu" "-C" "prefer-dynamic" "-o" "/home/matthias/vcs/github/rust-clippy/target/debug/test_build_base/used_underscore_binding_macro.stage-id" "-L" "target/debug" "-L" "target/debug/deps" "-Dwarnings" "-L" "/home/matthias/vcs/github/rust-clippy/target/debug/test_build_base/used_underscore_binding_macro.stage-id.aux"
stdout:
------------------------------------------
------------------------------------------
stderr:
------------------------------------------
error[E0464]: multiple matching crates for `serde`
  --> tests/run-pass/used_underscore_binding_macro.rs:10:10
   |
10 | #[derive(Deserialize)]
   |          ^^^^^^^^^^^
   |
   = note: candidates:
           crate `serde`: /home/matthias/vcs/github/rust-clippy/target/debug/deps/libserde-13a2511189973f33.rlib
error[E0463]: can't find crate for `_serde`
  --> tests/run-pass/used_underscore_binding_macro.rs:10:10
   |
10 | #[derive(Deserialize)]
   |          ^^^^^^^^^^^ can't find crate
error: aborting due to 2 previous errors
Some errors occurred: E0463, E0464.
For more information about an error, try `rustc --explain E0463`.
------------------------------------------
thread '[run-pass] run-pass/used_underscore_binding_macro.rs' panicked at 'explicit panic', /home/matthias/.cargo/registry/src/github.com-1ecc6299db9ec823/compiletest_rs-0.3.11/src/runtest.rs:2544:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
failures:
    [run-pass] run-pass/used_underscore_binding_macro.rs
test result: FAILED. 27 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
test compile_test ... FAILED
failures:
---- compile_test stdout ----
thread 'compile_test' panicked at 'Some tests failed', /home/matthias/.cargo/registry/src/github.com-1ecc6299db9ec823/compiletest_rs-0.3.11/src/lib.rs:89:22
failures:
    compile_test
test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
error: test failed, to rerun pass '--test compile-test'
````
If I just run  cargo clean ; cargo test though, there is no panic.

### PR DESCRIPTION
**Pull request by matthiaskrgr**
Fixes #2896
Fixes #2139

### CODE REVIEW

File: tests/compile-test.rs [hunk 1]
```diff
@@ -75,7 +75,10 @@ fn config(mode: &str, dir: PathBuf) -> compiletest::Config {
 }
 
 fn run_mode(mode: &str, dir: PathBuf) {
-    compiletest::run_tests(&config(mode, dir));
+    let cfg = config(mode,dir);
```
Comment (flip1995) [inline tests/compile-test.rs]: NIT: whitespace after `,`

**Comment by phansch** [comment]
I think this also fixes https://github.com/rust-lang-nursery/rust-clippy/issues/2139?

**Comment by matthiaskrgr** [comment]
Yup, looks like it.

**Comment by phansch** [comment]
Nice, that should make using RLS and Clippy together a much better experience  :fireworks: :sparkler: :sparkles:

**Pull request merged**